Native TBB parallel backend

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20480 Native TBB parallel backend**
* #20454 Restore TBB module
* #20157 Support for Eigen thread pool
* #20087 Native ATen/Parallel backend
* #20057 Split ATen/Parallel into interface and backend
* #20050 Move inter-op settings into ATen/Parallel

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB BUILD_BINARY=1
python setup.py develop --cmake

./build/bin/parallel_info
OMP_NUM_THREADS=5 ./build/bin/parallel_info
OMP_NUM_THREADS=5 MKL_NUM_THREADS=6 ./build/bin/parallel_info
./build/bin/thread_init_test
./build/bin/intra_inter_benchmark
./build/bin/intra_inter_benchmark --intra_op_threads 1
python  test/test_jit.py
python  test/test_torch.py

Differential Revision: [D15333692](https://our.internmc.facebook.com/intern/diff/D15333692)

Commit:
Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

cpuhrsch commented on aten/src/ATen/ParallelNativeTBB.cpp:
```diff
@@ -0,0 +1,75 @@
+namespace at {
+
+namespace internal {
+  tbb::task_scheduler_init tbb_init(
```
This is the trickiest part. It must be called before *anything* calls into TBB. I'm not even sure how nicely this plays with other packages that might use TBB. This might be one of the main reasons we need a static build for TBB, so that we can be sure this is correct. I also don't know what happens when the system uses two versions of TBB at the same time, i.e. if they can play nicely with each other.

ilia-cher replied:
As I understand, static build is not going to work with TBB used by PyTorch, MKL and MKL-DNN. As of now we compile MKL-DNN statically into PyTorch, so they could use statically compiled TBB, but it seems this is not going to work with dynamically loaded MKL

cpuhrsch commented on aten/src/ATen/ParallelNativeTBB.h:
```diff
@@ -0,0 +1,82 @@
+  } else {
+    tbb::task_group tg;
+    internal::_get_arena().execute([&](){
+      tg.run([&]() {
```
Maybe it makes sense to explicitly list the captured variables.

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

cpuhrsch commented on aten/src/ATen/ParallelNativeTBB.cpp:
```diff
@@ -0,0 +1,75 @@
+namespace internal {
+  tbb::task_scheduler_init tbb_init(
+      tbb::task_scheduler_init::deferred);
+  tbb::task_arena arena;
```
@colesbury - do you remember what the issue / if there was an issue with the approach of having a global arena for tbb control (to set the correct number of threads)?

colesbury replied:
No, I don't remember much other than something like having to set the number of threads before the first TBB launch

cpuhrsch replied:
This arena setup should get rid of that. However, maybe the access needs to be made atomic @ilia-cher?

ilia-cher replied:
as discussed offline I think I'll have to remove arena from here for perf reason

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

dzhulgakov commented on aten/src/ATen/ParallelNativeTBB.cpp:
```diff
@@ -0,0 +1,72 @@
+}
+
+int get_thread_num() {
+  auto index = internal::arena.current_thread_index();
```
according to https://software.intel.com/en-us/node/506359 it's better to call this_task_arena::current_thread_index

and why do you return 0 if index < 0?

ilia-cher replied:
I'll send an update, most likely I'll need to get rid of arena setup because it's slow

ilia-cher replied:
re. -1 and 0 - I'm trying to mimic omp behavior

ilia-cher replied:
but actually it might be better to TORCH_CHECK instead

dzhulgakov commented on aten/src/ATen/ParallelNativeTBB.h:
```diff
@@ -0,0 +1,104 @@
+
+  if ((end - begin) < grain_size || get_num_threads() == 1) {
+    return f(begin, end, ident);
+  } else {
```
nit: you  can drop `else` as there's return above

dzhulgakov commented on aten/src/ATen/ParallelNativeTBB.h:
```diff
@@ -0,0 +1,104 @@
+
+  if ((end - begin) < grain_size || get_num_threads() == 1) {
+    f(begin, end);
+  } else {
```
nit: you  can drop `else` and put return above

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

dzhulgakov commented on binaries/intra_inter_benchmark.cc:
```diff
@@ -0,0 +1,104 @@
+int main(int argc, char** argv) {
+  at::init_num_threads();
+
+  if (argc > 1) {
```
there are gflags (or a stub of them) in c10 (c10/util/Flags.h)

dzhulgakov reviewed (commented):
Looks good to me, I'll let @cpuhrsch  to do a pass too

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

cpuhrsch commented on aten/src/ATen/ParallelCommon.cpp:
```diff
@@ -18,9 +19,25 @@ namespace at {
+  } catch (const std::exception& e) {
+    std::ostringstream oss;
+    oss << "Invalid " << var_name << " variable value, " << e.what();
+    TORCH_WARN(oss.str());
```
I don't now if there is some aten specific prefix. Also, you don't see a lot of warnings in general as far as I know.

ilia-cher replied:
we're moving away from AT_ macro to TORCH_ atm

cpuhrsch commented on aten/src/ATen/ParallelCommon.cpp:
```diff
@@ -47,10 +64,31 @@ std::string get_parallel_info() {
+  ss << "\tMKL_NUM_THREADS : "
+     << get_env_var("MKL_NUM_THREADS", "[not set]") << std::endl;
+
+  ss << "Parallel backend: ";
```
Maybe instead of doing an if elif you could print all possibilities and then which one was chosen. We could have multiple switched on at once right?

ilia-cher replied:
no, the ones below are mutually exclusive, unfortunately
I'll print all possibilities

cpuhrsch commented on binaries/intra_inter_benchmark.cc:
```diff
@@ -0,0 +1,96 @@
+  typedef std::chrono::milliseconds ms;
+  std::chrono::time_point<clock> start_time = clock::now();
+
+  launch_tasks_and_wait(left, right, FLAGS_iter);
```
Might be good to run this multiple times to see how consistent it is.

cpuhrsch replied:
Multiple times within a benchmark

cpuhrsch commented on binaries/intra_inter_benchmark.cc:
```diff
@@ -0,0 +1,96 @@
+  auto right = at::ones({FLAGS_tensor_dim, FLAGS_tensor_dim}, at::kFloat);
+
+  // warmup
+  launch_tasks_and_wait(left, right, FLAGS_warmup_iter);
```
The warmup should actually be interesting in itself if there's some initialization going on.

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head

Commit:
Update on "Native TBB parallel backend"

Native TBB parallel backend

Summary:
Using TBB for explicit parallelization within ops

Test Plan:
PARALLEL_BACKEND=NATIVE_TBB USE_CUDA=0 USE_OPENMP=0 USE_TBB=1 MKL_TBB=1
BLAS=MKL USE_MKLDNN=1 MKLDNN_THREADING=TBB USE_EIGEN_THREADPOOL=1
BUILD_BINARY=1 python setup.py develop --cmake

./build/bin/parallel_info
tests

gh-metadata: pytorch pytorch 20480 gh/ilia-cher/19/head
